File: core/MinimumGeneticMutation/MinimumGeneticMutation.py
from typing import List
from collections import deque


def minMutation(start: str, end: str, bank: List[str]) -> int:
    if start == end:
        return 0
    bank = set(bank)
    if end not in bank:
        return -1
    queue = deque([(start, 0)])
    while queue:
        cur, step = queue.popleft()
        for i, x in enumerate(cur):
            for y in "ACGT":
                if y != x:
                    nxt = cur[:i] + y + cur[i+1:]
                    if nxt in bank:
                        if nxt == end:
                            return step + 1
                        bank.remove(nxt)
                        queue.append((nxt, step + 1))

    return -1

File: core/MinimumGeneticMutation/MinimumGeneticMutation.java
package com.dbc;

import javafx.util.Pair;

import java.util.*;

public class MinimumGeneticMutation {
    public int minMutation(String start, String end, String[] bank) {
        if (start.equals(end)) return 0;
        Set<String> set = new HashSet<>(Arrays.asList(bank));
        if (!set.contains(end)) return -1;
        char[] tran = new char[]{'A', 'C', 'G', 'T'};
        Deque<Pair<String, Integer>> queue = new ArrayDeque<>();
        queue.offerFirst(new Pair<>(start, 0));
        while (!queue.isEmpty()) {
            Pair<String, Integer> pair = queue.pollFirst();
            for (int i = 0; i < pair.getKey().length(); i++) {
                for (char ch : tran) {
                    if (ch != pair.getKey().charAt(i)) {
                        String temp = pair.getKey().substring(0, i) + ch + pair.getKey().substring(i + 1);
                        if (set.contains(temp)) {
                            if (temp.equals(end)) return pair.getValue() + 1;
                            set.remove(temp);
                            queue.addLast(new Pair<>(temp, pair.getValue() + 1));
                        }
                    }
                }
            }
        }
        return -1;
    }
}
